crates/cli/src/wasm_generator/static.rs

```diff
@@ -36,10 +39,94 @@ pub fn refine(mut wasm: Vec<u8>, js: &JS) -> Result<Vec<u8>> {
     module
         .run_optimization_passes(vec!["strip"], &codegen_cfg)
         .map_err(|_| anyhow!("Running wasm-opt optimization passes failed"))?;
-    wasm = module.write();
+    let wasm = module.write();
 
     let mut module = transform::module_config().parse(&wasm)?;
     module.customs.add(SourceCodeSection::new(js)?);
     transform::add_producers_section(&mut module.producers);
     Ok(module.emit_wasm())
 }
+
+fn export_exported_js_functions(wasm: &[u8], js_exports: Vec<String>) -> Result<Vec<u8>> {
+    let mut module = transform::module_config().parse(wasm)?;
+
+    let Exports {
+        realloc_export,
+        realloc_fn,
+        invoke_export,
+        invoke_fn,
+        memory,
+    } = get_useful_exports(&module);
+
+    // We no longer need these exports so remove them.
+    module.exports.delete(realloc_export);
+    module.exports.delete(invoke_export);
+
+    let ptr_local = module.locals.add(ValType::I32);
+    for js_export in js_exports {
+        // For each JS function export, add an export that copies the name of the function into memory and invokes it.
+        let js_export_bytes = js_export.as_bytes();
+        let js_export_len: i32 = js_export_bytes.len().try_into().unwrap();
+        let fn_name_data = module.data.add(DataKind::Passive, js_export_bytes.to_vec());
+
+        let mut export_fn = FunctionBuilder::new(&mut module.types, &[], &[]);
+        export_fn
+            .func_body()
+            .i32_const(0) // orig ptr
+            .i32_const(0) // orig len
+            .i32_const(1) // alignment
+            .i32_const(js_export_len) // new size
+            .call(realloc_fn)
+            .local_tee(ptr_local)
+            .i32_const(0) // offset into data segment
+            .i32_const(js_export_len) // size to copy
+            .memory_init(memory, fn_name_data) // copy fn name into allocated memory
+            .data_drop(fn_name_data)
+            .local_get(ptr_local)
+            .i32_const(js_export_len)
+            .call(invoke_fn);
+        let export_fn = export_fn.finish(vec![], &mut module.funcs);
+        module.exports.add(&js_export, export_fn);
+    }
+    let finished_wasm = module.emit_wasm();
+
+    Ok(finished_wasm)
+}
+
+struct Exports {
+    realloc_export: ExportId,
+    realloc_fn: FunctionId,
+    invoke_export: ExportId,
+    invoke_fn: FunctionId,
+    memory: MemoryId,
+}
+
+fn get_useful_exports(module: &walrus::Module) -> Exports {
```

Comment:
I feel like there's got to a better pattern to do what I'm trying to do in here. Namely, given a collection of things, find the ids for specific things that _should_ be in the collection.

Reply:
> I thought that Walrus would provide an api to achieve this easier, it seems that something like ModuleExports::by_name is what we need, similar to ModuleFunctions::by_name.

Yeah that would've been nice.

I did consider using `ModuleFunctions::by_name` but it's worth pointing out that the semantics with that method is different since it uses the `name` custom section and multiple functions can have the same name in that section and we don't control the name used.

What I could do is try to build a `Map<&str, &Export>` which would roughly correspond to how a `by_name` method would work, but I'd still need a pass to pull the underlying IDs out given the `ExportItem` variant on the export. I can experiment a little.

> can we rename this function to something else?

Yeah I was struggling with the name. I didn't really like `get_exports` because it's a subset of the exports. I didn't like `get_realloc_and_invoke_and_memory_ids`, especially given it would have to change depending on what's being returned. I'll experiment a little more but worst case, I'll leave a comment.

Reply:
I completely changed how I was going about this. I built a `HashMap<&str, &Export>` and was able to use the relatively new `let-else` syntax to get the ids I needed out. There's probably still some room for improvement but it meant I could get rid of the helper function and struct entirely.